sigproc/src/ResponseShaper.cxx

```diff
@@ -0,0 +1,120 @@
+/**
+ * A generic filter to correct reponse functions 
+ * Note: the purpose of this filter is not to replace the perchannel response
+ * correction in the OmnibusSigProc, instead it's implemented for validation purpose
+ */
+
+#include "WireCellSigProc/ResponseShaper.h"
+#include "WireCellSigProc/Derivations.h"
+
+#include "WireCellAux/DftTools.h"
+
+#include "WireCellUtil/NamedFactory.h"
+#include "WireCellUtil/Response.h"
+#include "WireCellIface/IChannelResponse.h"
+#include "WireCellIface/IWaveform.h"
+
+#include <cmath>
+#include <complex>
+#include <iostream>
+#include <set>
+
+WIRECELL_FACTORY(perChannelShaper, WireCell::SigProc::ResponseShaper::OneChannelResponse, WireCell::IChannelFilter,
+                 WireCell::IConfigurable)
+
+using namespace WireCell::SigProc;
+using WireCell::Aux::DftTools::fwd_r2c;
+using WireCell::Aux::DftTools::inv_c2r;
+
+
+ResponseShaper::OneChannelResponse::OneChannelResponse(const std::string& anode, const std::string& noisedb)
+  : m_anode_tn(anode)
+  , m_noisedb_tn(noisedb)
+{
+}
+ResponseShaper::OneChannelResponse::~OneChannelResponse() {}
+
+void ResponseShaper::OneChannelResponse::configure(const WireCell::Configuration& cfg)
+{
+    m_anode_tn = get(cfg, "anode", m_anode_tn);
+    m_anode = Factory::find_tn<IAnodePlane>(m_anode_tn);
+    m_noisedb_tn = get(cfg, "noisedb", m_noisedb_tn);
+    m_noisedb = Factory::find_tn<IChannelNoiseDatabase>(m_noisedb_tn);
+
+    std::string dft_tn = get<std::string>(cfg, "dft", "FftwDFT");
+    m_dft = Factory::find_tn<IDFT>(dft_tn);
+}
+WireCell::Configuration ResponseShaper::OneChannelResponse::default_configuration() const
+{
+    Configuration cfg;
+    cfg["anode"] = m_anode_tn;
+    cfg["noisedb"] = m_noisedb_tn;
+    cfg["dft"] = "FftwDFT";     // type-name for the DFT to use
+    return cfg;
+}
+
+WireCell::Waveform::ChannelMaskMap ResponseShaper::OneChannelResponse::apply(int ch, signal_t& signal) const
+{
+    WireCell::Waveform::ChannelMaskMap ret;
+
+    // correct rc undershoot
+    auto spectrum = fwd_r2c(m_dft, signal);
+
+    // now apply the ch-by-ch response ...
+    auto cr = Factory::find_tn<IChannelResponse>("ParamsPerChannelResponse");
+    auto cr_bins = cr->channel_response_binning();
+    // if (cr_bins.binsize() != m_period) { // FIXME
+    //     THROW(ValueError() << errmsg{"ProtoduneHD::OneChannelResponse channel response size mismatch"});
+    // }
+    auto period = cr_bins.binsize();
+    WireCell::Binning tbins(signal.size(), cr_bins.min(), cr_bins.min() + signal.size() * period);
+    Response::ColdElec ce(14 * units::mV / units::fC, 2.2 * units::us);
```

Comment:
At the very least, these parameters should be configurable.  

But to make this component more general, not assuming just the CE response function, the configuration can name an `IWaveform`

Reply:
@brettviren Good point. I have now retrieved the elecreponse from configuration. Please take a look.

Reply:
Thanks @wenqiang-gu this is very good.